[#98109110] Make AWS NAT server run on ubuntu

mtekel opened the pull request:
Make the AWS NAT server run using ubuntu, just like the GCE NAT and all other hosts.

This is needed for shutdown scripts. Otherwise there's problem running ansible against the NAT box itself, as it can't ssh to itself, because it's using different user name than all the hosts.

Additionally, we found a problem in the NAT server which prevents it from setting up NAT routing after reboot, as it was not installing a package to persist the iptables state.

After this is merged, ssh config template needs updating: https://github.com/alphagov/tsuru-ansible/pull/113

mtekel committed:
Make AWS NAT server run on ubuntu

Make the AWS NAT server run on ubuntu, just like the GCE NAT and all other hosts.

mtekel committed:
Remove unused NAT ami variables

AWS NAT server runs on the same ubuntu image as all other hosts. Thus there's no need for the variable specifying different AMI for NAT server anymore.

dcarley commented:
Not necessarily a blocker for merging, but 70e15dc doesn't include any of this good detail about why we're doing this (which was the pressing reason for doing this rather than just consistency):

> This is needed for shutdown scripts. Otherwise there's problem running ansible against the NAT box itself, as it can't ssh to itself, because it's using different user name than all the hosts.
> 
> Additionally, we found a problem in the NAT server which prevents it from setting up NAT routing after reboot, as it was not installing a package to persist the iptables state.

dcarley commented on setup-nat-routing.sh:
```diff
@@ -0,0 +1,28 @@
+
+echo "Save iptables state"
+sudo apt-get update
+#sudo apt-get update | tee /tmp/second.txt # I need to do this twice!!! Why!?!?!?!
```
What's this? Is it needed?

I think it'd be good to break the move to a single script and the modifications required to make it run on the new AWS AMI into separate commits, because it's difficult to tell what's changed here and where it's come from.

mtekel commented:
There are essentially 2 reasons for doing this.
1. This looks like a technical debt. All the servers are on ubuntu, even google NAT, just this one server is special.
2. And now this server being special also breaks ansible shutdown. 

Regarding 70e15dc: If you look at it just as a single commit, you don't necessarily need to have full context. Therefore I only mention reason 1 in the commit message, because reason 2 is happening in the different repo. Reason 1 is IMO a good enough reason for fixing this, but I can of course always amend the commit message...

dcarley commented:
We're changing it right now because it breaks the Ansible shutdown, not because we were looking for tech debt to fix, so we should include that in the commit message. A commit should give enough detail to allow another person in the future to understand the implications (as understood at the time it was originally committed - not everything, ever) of reverting that commit if necessary.

dcarley commented on aws/nat-server.tf:
```diff
@@ -1,27 +1,18 @@
-      "echo 'net.ipv4.ip_forward = 1' | sudo tee /etc/sysctl.d/nat.conf",
-      "echo 'net.ipv4.conf.eth0.send_redirects = 0' | sudo tee -a /etc/sysctl.d/nat.conf"
-    ]
+    script = "../setup-nat-routing.sh"
```
I don't necessarily think that this should be reverted, but just to note, that changing this to a script prevents any output from being shown. So we don't see the additional `echo` statements that have been added and it's hard to debug whether it's doing anything when this happens:

```
aws_instance.nat: Provisioning with 'remote-exec'...
aws_instance.nat (remote-exec): Connecting to remote host via SSH...
aws_instance.nat (remote-exec):   Host: 52.17.243.1
aws_instance.nat (remote-exec):   User: ubuntu
aws_instance.nat (remote-exec):   Password: false
aws_instance.nat (remote-exec):   Private key: true
aws_instance.nat (remote-exec):   SSH Agent: false
aws_instance.nat (remote-exec): Connecting to remote host via SSH...
aws_instance.nat (remote-exec):   Host: 52.17.243.1
aws_instance.nat (remote-exec):   User: ubuntu
aws_instance.nat (remote-exec):   Password: false
aws_instance.nat (remote-exec):   Private key: true
aws_instance.nat (remote-exec):   SSH Agent: false
aws_instance.nat (remote-exec): Connecting to remote host via SSH...
aws_instance.nat (remote-exec):   Host: 52.17.243.1
aws_instance.nat (remote-exec):   User: ubuntu
aws_instance.nat (remote-exec):   Password: false
aws_instance.nat (remote-exec):   Private key: true
aws_instance.nat (remote-exec):   SSH Agent: false
aws_instance.nat (remote-exec): Connecting to remote host via SSH...
aws_instance.nat (remote-exec):   Host: 52.17.243.1
aws_instance.nat (remote-exec):   User: ubuntu
aws_instance.nat (remote-exec):   Password: false
aws_instance.nat (remote-exec):   Private key: true
aws_instance.nat (remote-exec):   SSH Agent: false
aws_instance.nat (remote-exec): Connected!
```

… several minutes pass …

```
aws_instance.nat: Error: 1 error(s) occurred:

* Script exited with non-zero exit status: 100
```

NB: If we are able to fix the output again, then we should probably quieten down the `apt-get` operations as they were before.

keymon committed:
Move provisioning scripts of NAT server to script

The inline commands in terraform will not stop the execution if
they fail, you have to use an external script for that.
Additionally, an external script would be easier to edit.

mtekel commented:
Increasing output verbosity was intentional as we were debugging some issues with apt-get update. The output is now being stored to `/tmp/setup-nat-routing.log`. This makes debugging easier, esp. when you need to access the host some time after it was created and you want to check logs.

keymon commented:
I can change the commit to move the removal of -qq to the next one

keymon committed:
Wait for cloudinit in NAT server provision script
    
During the NAT server provisioning we need to install some
using apt-get. But cloudinit might be running at that mome
need it to update the upstream repositories.

keymon committed:
Redirect output in NAT server provision
    
We redirect the output in the provision script of the NAT 
for better troubleshooting. 
Also add some logging messages and increase the verbosity
of apt-get commands.

dcarley commented:
Thanks. That's much easier to understand now :+1:
